### PR TITLE
[CLI] Initial implementation for batch generation

### PR DESCRIPTION
**Pull request by jimschubert**
### PR checklist

- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [ ] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

> *THIS IS A WIP* Requesting feedback before I convert all exiting JSON based configs which are executed by `ensure-up-to-date`

Allows for generating multiple outputs via config. Just specify multiple config files on command line.

Intent for this is to reduce CI times to generate outputs as well as to reduce time for users to run ensure-up-to-date to meet PR standards.

Example command:

```bash
openapi-generator batch --includes-base-dir `pwd` --fail-fast  -- bin/ci/*
```

---

As part of this implementation, the batch command support a customized JSON key, `!include`. If this key's value refers to an existing file, that file's contents are "unwrapped" into the config during deserialization. This allows us to easily point to the same configs used by our sample scripts without modifying the CLI generate task's switches or assumptions. 

cc @OpenAPITools/generator-core-team

### CODE REVIEW

**Comment by jimschubert** [comment]
If it's ever needed again… here's the small program I used to generate JSON from existing bash scripts: https://github.com/jimschubert/openapi-generator-commands-processor

File: modules/openapi-generator-cli/src/main/java/org/openapitools/codegen/cmd/GenerateBatch.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+            LOGGER.error("No configuration file inputs specified");
+            System.exit(1);
+        }
+
+        int cores = Runtime.getRuntime().availableProcessors();
+        int numThreads = 2 * cores;
+        if (null != threads && (threads > 0 && threads < Thread.activeCount())) {
+            numThreads = threads;
+        }
+
```
Comment (wing328) [inline modules/openapi-generator-cli/src/main/java/org/openapitools/codegen/cmd/GenerateBatch.java]: Minor suggestion: maybe good to use LOGGER.info to show how many threads are used as an FYI to the user.

**Comment by wing328** [comment]
I believe you will update the doc (https://openapi-generator.tech/docs/usage) later as part of the upcoming release

**Comment by wing328** [comment]
@jimschubert Overall it looks pretty good 👍Did simple tests locally and worked as expected. The CircleCI test failed with a weird error I've never seen. I"ve just restarted the job to see if the issue goes away.

**Comment by wing328** [comment]
Some Petstore Play server files have changed ...
```
Perform git status
On branch batch-generation
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   samples/server/petstore/java-play-framework-async/app/controllers/PetApiController.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/PetApiControllerImp.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/PetApiControllerImpInterface.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/StoreApiController.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/StoreApiControllerImp.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/StoreApiControllerImpInterface.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/UserApiController.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/UserApiControllerImp.java
	modified:   samples/server/petstore/java-play-framework-async/app/controllers/UserApiControllerImpInterface.java
````

Is that due to mismatch in the script (play server) and the corresponding JSON files under ./bin/ci ?

**Comment by jimschubert** [comment]
@wing328 that could be due to CI running against the merge commit. I'd need to merge master into this branch and regenerate samples to verify.

**Comment by jimschubert** [comment]
Regarding doc update, if this looks good and we decide to use it... I'll include doc update on this PR.

**Comment by wing328** [comment]
> that could be due to CI running against the merge commit.

@jimschubert I don't think CircleCI does that but I could be wrong. From what I know only shippable and Appveyor will try to test against the merge commit.

Anyway, it's still good to merge the latest master into this branch. I'll do it later today to see if the issue goes away.

**Comment by jimschubert** [comment]
> I believe you will update the doc (https://openapi-generator.tech/docs/usage) later as part of the upcoming release

@wing328 the usage page is all "public" commands. The batch generation command is hidden, and I think it should remain CI-only until we work out any kinks.  For example, I'm wondering if we should change all shell scripts directly under `./bin` to use externalized configs. If we did this, we could merge the `./bin/ci/*.json` contents with the relevant configs under `./bin/*.json` (excluding the `!include` customization). The only problem there would be in cases where `./bin/*.json` files are shared across sample outputs for generators with different options applied, such as the ruby client.

I'm thinking about incorporating this into the CI scripts as a separate PR. This will require that we run the `./bin/meta-codegen.sh` script separately.

**Pull request merged**